src/ripple/app/reporting/ETLHelpers.h

```diff
@@ -0,0 +1,194 @@
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_ETLHELPERS_H_INCLUDED
+#define RIPPLE_CORE_ETLHELPERS_H_INCLUDED
+#include <ripple/app/main/Application.h>
+#include <ripple/ledger/ReadView.h>
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <sstream>
+
+namespace ripple {
+
+/// This datastructure is used to keep track of the sequence of the most recent
+/// ledger validated by the network. There are two methods that will wait until
+/// certain conditions are met. This datastructure is able to be "stopped". When
+/// the datastructure is stopped, any threads currently waiting are unblocked.
+/// Any later calls to methods of this datastructure will not wait. Once the
+/// datastructure is stopped, the datastructure remains stopped for the rest of
+/// it's lifetime.
+class NetworkValidatedLedgers
+{
+    // max sequence validated by network
+    std::optional<uint32_t> max_;
+
+    std::mutex mtx_;
+
+    std::condition_variable cv_;
+
+    bool stopping_ = false;
+
+public:
+    /// Notify the datastructure that idx has been validated by the network
+    /// @param idx sequence validated by network
+    void
+    push(uint32_t idx)
+    {
+        std::unique_lock<std::mutex> lck(mtx_);
+        if (!max_ || idx > *max_)
+            max_ = idx;
+        cv_.notify_all();
+    }
+
+    /// Get most recently validated sequence. If no ledgers are known to have
+    /// been validated, this function waits until the next ledger is validated
+    /// @return sequence of most recently validated ledger. empty optional if
+    /// the datastructure has been stopped
+    std::optional<uint32_t>
+    getMostRecent()
+    {
+        std::unique_lock<std::mutex> lck(mtx_);
+        cv_.wait(lck, [this]() { return max_ || stopping_; });
+        return max_;
+    }
+
+    /// Waits for the sequence to be validated by the network
+    /// @param sequence to wait for
+    /// @return true if sequence was validated, false otherwise
+    /// a return value of false means the datastructure has been stopped
+    bool
+    waitUntilValidatedByNetwork(uint32_t sequence)
+    {
+        std::unique_lock<std::mutex> lck(mtx_);
+        cv_.wait(lck, [sequence, this]() {
+            return (max_ && sequence <= *max_) || stopping_;
+        });
+        return !stopping_;
+    }
+
+    /// Puts the datastructure in the stopped state
+    /// Future calls to this datastructure will not block
+    /// This operation cannot be reversed
+    void
+    stop()
+    {
+        std::unique_lock<std::mutex> lck(mtx_);
```

Comment:
In this place, `std::lock_guard` is better because it can't let go of the lock until destruction.

Reply:
Can't pass `std::lock_guard` to `std::condition_variable::wait()`: https://en.cppreference.com/w/cpp/thread/condition_variable/wait

Reply:
sorry, in this place we actually can use `std::lock_guard` because we aren't passing it to `wait`

Reply:
changed to `std::lock_guard`